### PR TITLE
Add Levenberg-Marquardt model calibration

### PR DESCRIPTION
**Pull request by richterjakob**

### CODE REVIEW

**Comment by mrp089** [comment]
Thank you so much, @richterjakob! I will have a look with @menon-karthik. This pull request closes issue #24.

File: src/model/bloodvessel.hpp [hunk 1]
```diff
@@ -110,40 +103,51 @@ namespace MODEL {
...
+ *
+ * ### Parameters
+ *
+ * Parameter sequence for constructing this block
+ *
+ * * `0` Poiseuille resistance
+ * * `1` Capacitance
+ * * `2` Inductance
+ * * `3` Stenosis coefficient
+ *
```
Comment (mrp089) [inline src/model/bloodvessel.hpp:132]: Thanks for adding this to the documentation!

File: src/optimize/levenbergmarquardtoptimizer.hpp [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+ * }}^\alpha, \f]
+ *
+ * The Jacobian is derived from the residual as
+ *
+ * \f[
+ * J = \frac{\partial \boldsymbol{r}}{\partial \boldsymbol{\alpha}} =
+ * \frac{\partial \mathbf{E}}{\partial \boldsymbol{\alpha}} \cdot
+ * \dot{\mathbf{y}}+\frac{\partial \mathbf{F}}{\partial \boldsymbol{\alpha}}
+ * \cdot \mathbf{y}+\frac{\partial \mathbf{c}}{\partial \boldsymbol{\alpha}} \f]
+ *
```
Comment (mrp089) [inline src/optimize/levenbergmarquardtoptimizer.hpp:102]: Looks great!

File: applications/svzerodcalibrator.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  DEBUG_MSG("Parse configuration");
+  std::ifstream ifs(input_file);
+  const auto& config = nlohmann::json::parse(ifs);
+
+  auto output_config = OPT::calibrate<T>(config);
+
+  // Write optimized simulation config
+  std::ofstream o(output_file);
+  o << std::setw(4) << output_config << std::endl;
+}
```
Comment (mrp089) [inline applications/svzerodcalibrator.cpp:64]: Nice idea of having a separate solver and calibrator!

**Review by mrp089 — CHANGES_REQUESTED** [review]
Awesome work!! I'm super happy with everything calibrator. @menon-karthik will look at how the code changes impact his interface in the next couple of days. I would only change the test case setup (see comments). Otherwise, good to go!

**Comment by menon-karthik** [comment]
Update: I think I got the interface and coupling to svSolver to work now. I have also created a simple C++ test case for the interface while doing this, which should be added to the github actions (#30 ). Still wondering what would be the best way to add a test case specifically for the svSolver (and eventually svFSIPlus) coupling. I will push my changes in the next few days!

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -8,20 +8,14 @@ jobs:
...
         if: ${{ matrix.os == 'ubuntu-*' }}
         run: sudo apt update && sudo apt install build-essential cmake
-      - name: Build svZeroDSolver
-        run: |
-          mkdir Release
-          cd Release
-          cmake -DCMAKE_BUILD_TYPE=Release ..
-          cmake --build .
-      - name: Install test dependencies
+      - name: Install svZeroDPlus
```
Comment (menon-karthik) [inline .github/workflows/test.yml:18]: I'm having trouble running pytest when I build using CMake on Sherlock. The problem is that pytest cannot find the svzerodplus module when importing it in test_integration_cpp.py. Suggestions?

**Review by menon-karthik — COMMENTED** [review]
Looks great to me for the most part! The only problem is pytest not working when using the CMake build on Sherlock.

File: .github/workflows/test.yml (hunk 4 above)
Comment (richterjakob) [reply .github/workflows/test.yml:18]: See below

**Comment by mrp089** [comment]
Thanks for adding the test, @richterjakob! I'll move the `cali.py` script to a repository for the upcoming paper.

**Comment by richterjakob** [comment]
> Thanks for adding the test, @richterjakob! I'll move the `cali.py` script to a repository for the upcoming paper.

Haha, you were faster than I could type that the only thing left is to decide whether the keeping `cali.py` in this repository is the right choice 😄

**Comment by richterjakob** [comment]
But otherwise I think we are ready to merge, right?

**Comment by menon-karthik** [comment]
> > One issue I'm having is running the test cases when I build the code using CMake instead of pip (on Sherlock). In this case, pytest cannot find the svzerodplus package that is required to run tests/test_integration_cpp.py. Is there a way to have this work when using the CMake installation?
> 
> Unfortunately the python interface (which pytest is using) only works when installing svZeroDPlus via pip. But the pip installation should work on brocken.

In that case, one last thing I need to do is test the pip installation and purest on Sherlock. If that doesn't work, t's not huge deal because people pushing from Sherlock can still use Github actions in their own branches to run the tests. But I will add a note in the documentation if that is the case. I should have this done this morning.

If there's nothing else to add (apart from moving the `cali.py` script), I can merge after I test that. Does that sound good @richterjakob and @mrp089 ?

**Comment by richterjakob** [comment]
> > > One issue I'm having is running the test cases when I build the code using CMake instead of pip (on Sherlock). In this case, pytest cannot find the svzerodplus package that is required to run tests/test_integration_cpp.py. Is there a way to have this work when using the CMake installation?
> > 
> > 
> > Unfortunately the python interface (which pytest is using) only works when installing svZeroDPlus via pip. But the pip installation should work on brocken.
> 
> In that case, one last thing I need to do is test the pip installation and pytest on Sherlock. If that doesn't work, it's not huge deal because people pushing from Sherlock can still use Github actions in their own branches to run the tests. But I will add a note in the documentation if that is the case. I should have this done this morning.
> 
> If there's nothing else to add (apart from moving the `cali.py` script), I can merge after I test that. Does that sound good @richterjakob and @mrp089 ?

I tested the build on Sherlock already and it worked without problems. 

If that's okay, I would like to do the honour of merging this 😁

**Review by mrp089 — APPROVED** [review]

**Pull request merged**